src/schemes/fluid/implicit_incompressible_sph/system.jl

```diff
@@ -395,48 +379,135 @@ function calculate_diagonal_elements!(a_ii, system, neighbor_system::AbstractBou
     end
 end
 
+function calculate_predicted_density!(system, v, u, v_ode, u_ode, semi, t)
+    return system
+end
+
+function calculate_predicted_density!(system::ImplicitIncompressibleSPHSystem, v, u, v_ode,
+                                      u_ode, semi, t)
+    (; density, predicted_density, time_step) = system
+
+    predicted_density .= density
+
+    # Calculate the predicted density (with the continuity equation and predicted velocities)
+    foreach_system(semi) do neighbor_system
+        u_neighbor_system = wrap_u(u_ode, neighbor_system, semi)
+        system_coords = current_coordinates(u, system)
+        neighbor_system_coords = current_coordinates(u_neighbor_system, neighbor_system)
+
+        foreach_point_neighbor(system, neighbor_system, system_coords,
+                               neighbor_system_coords, semi,
+                               points=each_integrated_particle(system)) do particle,
+                                                                           neighbor,
+                                                                           pos_diff,
+                                                                           distance
+            # Calculate the predicted velocity differences
+            advection_velocity_diff = predicted_velocity(system, particle) -
+                                      predicted_velocity(neighbor_system, neighbor)
+            m_b = hydrodynamic_mass(neighbor_system, neighbor)
+            grad_kernel = smoothing_kernel_grad(system, pos_diff, distance, particle)
+            # Compute \rho_adv in eq. 4 in Ihmsen et al. (2013)
+            predicted_density[particle] += time_step * m_b *
+                                           dot(advection_velocity_diff, grad_kernel)
+        end
+    end
+end
+
 # Calculate pressure values with iterative pressure solver (relaxed Jacobi scheme)
-function pressure_solve!(system, v, u, v_ode, u_ode, semi, t)
-    (; reference_density, max_error, min_iterations, max_iterations, time_step) = system
+function pressure_solve!(semi, v_ode, u_ode, t)
+    foreach_system(semi) do system
+        initialize_pressure!(system, semi)
+    end
+
+    # Determine global iteration and error constraints across all iisph systems
+    min_iters = 1
+    max_iters = 10000
+    max_err = 100.0
+
+    foreach_system(semi) do system
+        if system isa ImplicitIncompressibleSPHSystem
+            (; max_error, min_iterations, max_iterations) = system
+
+            min_iters = max(min_iterations, min_iters)
+            max_iters = min(max_iterations, max_iters)
+            max_err = min(max_error, max_err)
+        end
+    end
```

Comment:
This is the source for your allocations. It is a type instability because the variables are defined outside of the `foreach_system` and then must be boxed to be used inside (the same problem that you solved with `Ref` for the average density error).

I guess this would be a more elegant solution:

```suggestion
    # Determine global iteration and error constraints across all IISPH systems
    min_iters = maximum(minimum_iisph_iterations, semi.systems)
    max_iters = minimum(maximum_iisph_iterations, semi.systems)
    max_err = minimum(maximum_iisph_error, semi.systems)
```

And then define these functions above:
```julia
@inline maximum_iisph_error(system) = convert(eltype(system), Inf)

@inline function maximum_iisph_error(system::ImplicitIncompressibleSPHSystem)
    return system.max_error
end

@inline minimum_iisph_iterations(system) = 0

@inline function minimum_iisph_iterations(system::ImplicitIncompressibleSPHSystem)
    return system.min_iterations
end

@inline maximum_iisph_iterations(system) = typemax(Int)

@inline function maximum_iisph_iterations(system::ImplicitIncompressibleSPHSystem)
    return system.max_iterations
end
```